services/src/transaction_replayer.rs

```diff
@@ -45,56 +42,44 @@ impl TransactionReplayer {
+                if Instant::now() < tx_replay.replay_at {
+                    tokio::time::sleep_until(tx_replay.replay_at).await;
                 }
-                let tx = reciever.recv().await;
-                match tx {
-                    Some(mut tx_replay) => {
-                        MESSAGES_IN_REPLAY_QUEUE.dec();
-                        if Instant::now() < tx_replay.replay_at {
-                            tokio::time::sleep_until(tx_replay.replay_at).await;
-                        }
-                        if let Some(tx) = tx_store.get(&tx_replay.signature) {
-                            if tx.status.is_some() {
-                                // transaction has been confirmed / no retry needed
-                                continue;
-                            }
-                        } else {
-                            // transaction timed out
-                            continue;
-                        }
-                        // ignore reset error
-                        let _ = tpu_service
-                            .send_transaction(tx_replay.signature.clone(), tx_replay.tx.clone());
-
-                        if tx_replay.replay_count < tx_replay.max_replay {
-                            tx_replay.replay_count += 1;
-                            tx_replay.replay_at = Instant::now() + retry_after;
-                            if let Err(e) = sender.send(tx_replay) {
-                                error!("error while scheduling replay ({})", e);
-                                continue;
-                            } else {
-                                MESSAGES_IN_REPLAY_QUEUE.inc();
-                            }
-                        }
+                if let Some(tx) = tx_store.get(&tx_replay.signature) {
+                    if tx.status.is_some() {
+                        // transaction has been confirmed / no retry needed
+                        continue;
                     }
-                    None => {
-                        error!("transaction replay channel broken");
-                        break;
+                } else {
+                    // transaction timed out
+                    continue;
+                }
+                // ignore reset error
+                let _ =
+                    tpu_service.send_transaction(tx_replay.signature.clone(), tx_replay.tx.clone());
+
+                if tx_replay.replay_count < tx_replay.max_replay {
+                    tx_replay.replay_count += 1;
+                    tx_replay.replay_at = Instant::now() + retry_after;
+                    if let Err(e) = sender.send(tx_replay) {
+                        error!("error while scheduling replay ({})", e);
+                        continue;
+                    } else {
+                        MESSAGES_IN_REPLAY_QUEUE.inc();
                     }
                 }
             }
-            Ok(())
+
+            bail!("transaction replay channel broken");
```

Comment:
here to add an explicit error! log

Reply:
same here. It will be logged down the line